front/lib/api/assistant/actions/retrieval.ts

```diff
@@ -244,6 +242,455 @@ export class RetrievalAction extends BaseAction {
+            chunks: {
+              text: string;
+              hash: string;
+              offset: number;
+              score: number;
+            }[];
+            token_count: number;
+          }[];
+
+          const refs = getRefs().slice(refsOffset, refsOffset + v.length);
+
+          documents = v.map((d, i) => {
+            const reference = refs[i % refs.length];
+            return {
+              id: 0, // dummy pending database insertion
+              dataSourceWorkspaceId:
+                dataSourcesIdToWorkspaceId[d.data_source_id],
+              dataSourceId: d.data_source_id,
+              documentId: d.document_id,
+              reference,
+              timestamp: d.timestamp,
+              tags: d.tags,
+              sourceUrl: d.source_url ?? null,
+              score: d.chunks.map((c) => c.score)[0],
+              chunks: d.chunks.map((c) => ({
+                text: c.text,
+                offset: c.offset,
+                score: c.score,
+              })),
+            };
+          });
+        }
+      }
+    }
+
+    // We are done, store documents and chunks in database and yield the final events.
+
+    await frontSequelize.transaction(async (t) => {
+      for (const d of documents) {
+        const document = await RetrievalDocument.create(
+          {
+            dataSourceWorkspaceId: d.dataSourceWorkspaceId,
+            dataSourceId: d.dataSourceId,
+            sourceUrl: d.sourceUrl,
+            documentId: d.documentId,
+            reference: d.reference,
+            documentTimestamp: new Date(d.timestamp),
+            tags: d.tags,
+            score: d.score,
+            retrievalActionId: action.id,
+          },
+          { transaction: t }
+        );
+
+        d.id = document.id;
+
+        for (const c of d.chunks) {
+          await RetrievalDocumentChunk.create(
+            {
+              text: c.text,
+              offset: c.offset,
+              score: c.score,
+              retrievalDocumentId: document.id,
+            },
+            { transaction: t }
+          );
+        }
+      }
+    });
+
+    logger.info(
+      {
+        workspaceId: conversation.owner.sId,
+        conversationId: conversation.sId,
+        elapsed: Date.now() - now,
+      },
+      "[ASSISTANT_TRACE] Retrieval action execution"
+    );
+
+    yield {
+      type: "retrieval_success",
+      created: Date.now(),
+      configurationId: agentConfiguration.sId,
+      messageId: agentMessage.sId,
+      action: new RetrievalAction({
+        id: action.id,
+        agentMessageId: agentMessage.agentMessageId,
+        params: {
+          relativeTimeFrame: relativeTimeFrame,
+          query: query,
+          topK,
+        },
+        functionCallId: action.functionCallId,
+        functionCallName: action.functionCallName,
+        documents,
+        step: action.step,
+      }),
+    };
+  }
+}
+
 export function retrievalAutoQueryInputSpecification() {
```

Comment:
No changes in this one and the following ones right? (asking cause I did not review them, assuming it's only moved code)

Reply:
Yep correct.